pandas/core/strings.py

```diff
@@ -502,6 +502,85 @@ def f(x):
                                dtype=object)
     return result, name
 
+def str_extractall(arr, pat, flags=0):
+    """
+    Find all non-overlapping groups in each 
+    subject string in the Series using regular expression pat.
+
+    Parameters
+    ----------
+    pat : string
+        Regular expression pattern
+    flags : int, default 0 (no flags)
+        re module flags, e.g. re.IGNORECASE
+
+    Returns
+    -------
+    A Series with one element for each match if there is only
+    one group in pat. Otherwise, a DataFrame with one row 
+    for each match and one column for each group.
+
+    Examples
+    --------
+    A pattern with one group will return a Series.
+    Indices with no matches will not appear in the result.
+
+    >>> S = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
+    >>> S.str.extractall("[ab](?P<digit>\d)")
+    A    1
+    A    2
+    B    1
+    Name: digit, dtype: object
+    
+    A pattern with multiple groups will return a DataFrame.
+    
+    >>> S.str.extractall("(?P<letter>[ab])(?P<digit>\d)")
+      letter digit
+    A      a     1
```

Comment:
not sure what you mean. Can you please clarify?

When the input Series is not multi-indexed, there is no reason the output DataFrame should be. This is the same as the behavior of the standard `extract` method:

```python
>>> from pandas import Series
>>> S = Series(["a1a2", "b1", "c1"], ["A", "B", "C"])
>>> S.str.extractall("(?P<letter>[ab])(?P<digit>\d)")
  letter digit
A      a     1
A      a     2
B      b     1
>>> S.str.extract("(?P<letter>[ab])(?P<digit>\d)")
  letter digit
A      a     1
B      b     1
C    NaN   NaN
>>> e_df = S.str.extract("(?P<letter>[ab])(?P<digit>\d)")
>>> e_df.index
Index([u'A', u'B', u'C'], dtype='object')
>>> e_df.keys()
Index([u'letter', u'digit'], dtype='object')
>>> ea_df = S.str.extractall("(?P<letter>[ab])(?P<digit>\d)")
>>> ea_df.index
Index([u'A', u'A', u'B'], dtype='object')
>>> ea_df.keys()
Index([u'letter', u'digit'], dtype='object')
>>> 
```